### PR TITLE
Issue 3114 event buffers for c

### LINKED ISSUE #3114: Add event buffers for C code generator
Is subtask of #3113 'Add event buffer for cycle based statecharts'

### PR DESCRIPTION
**Pull request by terfloth**
fixes #3114

### CODE REVIEW

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/StatemachineHeaderFragment.xtend [hunk 1]
```diff
@@ -184,7 +224,7 @@ class StatemachineHeaderFragment implements IHeaderFragment {
 	}	
 	
 	def dispatch scopeFunctionPrototypes(StatechartScope it) '''
-		«FOR d : declarations»
+		«FOR d : declarations.filter[d | !(d instanceof Event) || !((d as Event).isBufferEvent)]»
```
Comment (terfloth) [inline plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/StatemachineHeaderFragment.xtend:227]: Check if outdated

Comment (terfloth) [reply plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/StatemachineHeaderFragment.xtend:227]: Change not required anymore - fixed it

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/StatemachineHeaderFragment.xtend [hunk 2]
```diff
@@ -112,11 +119,44 @@ class StatemachineHeaderFragment implements IHeaderFragment {
...
+			«t.typeDeclaration(flow)»
+			«t.structDeclaration(flow)»
+		«ENDFOR»
+		
 		«statemachineStruct»
 		
 		'''
 	}
 	
+	def derivedComplexTypes(ExecutionFlow flow) {
```
Comment (terfloth) [inline plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/StatemachineHeaderFragment.xtend:133]: move to sexec

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/StatemachineSourceFragment.xtend [hunk 3]
```diff
@@ -107,11 +119,11 @@ class StatemachineSourceFragment implements ISourceFragment {
 		
 		«init»
 		
-		«enter»
+		«enterCode»
```
Comment (terfloth) [inline plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/StatemachineSourceFragment.xtend:122]: generate public methods generically

Comment (terfloth) [reply plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/StatemachineSourceFragment.xtend]: done

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/StatemachineHeaderFragment.xtend (hunk 2 above)
Comment (terfloth) [reply plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/StatemachineHeaderFragment.xtend]: done

File: plugins/org.yakindu.base.types/src/org/yakindu/base/types/annotations/VisibilityAnnotations.xtend [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package org.yakindu.base.types.annotations
```
Comment (terfloth) [inline plugins/org.yakindu.base.types/src/org/yakindu/base/types/annotations/VisibilityAnnotations.xtend:1]: Add File comment

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/CFlowConfiguration.xtend [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package org.yakindu.sct.generator.c
```
Comment (terfloth) [inline plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/CFlowConfiguration.xtend:1]: Add file comment

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/StatechartTypes.xtend [hunk 6]
```diff
@@ -151,8 +160,8 @@ class StatechartTypes {
 	def typeRelevantDeclarations(Scope scope){
 		return scope.declarations.filter[
 			switch it {
-				EventDefinition: true
-				TimeEvent: true
+				Event: true
+//				TimeEvent: true
```
Comment (terfloth) [inline plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/StatechartTypes.xtend:164]: remove line

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/DefaultNamingService.xtend [hunk 7]
```diff
@@ -268,6 +277,8 @@ class DefaultNamingService implements INamingService {
 	}
 
 	def protected prefix(Method it, char separator) {
+//		if (it.isPublic) flow.name.toFirstLower
```
Comment (terfloth) [inline plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/DefaultNamingService.xtend:280]: remove

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/tree/TreeNamingService.xtend [hunk 8]
```diff
@@ -305,7 +313,9 @@ class TreeNamingService implements INamingService {
 	}
 
 	def protected List<String> prefix(Method it) {
-		return new ArrayList<String>();
+		val l = new ArrayList<String>()
+//		if (it.isPublic) l.add(flow.name.toFirstLower)
```
Comment (terfloth) [inline plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/tree/TreeNamingService.xtend:317]: remove

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/transformation/ng/BehaviorMapping.xtend [hunk 9]
```diff
@@ -22,6 +23,7 @@ class BehaviorMapping extends org.yakindu.sct.model.sexec.transformation.Behavio
 	@Inject extension StatechartAnnotations
 	
 	@Inject extension ExpressionBuilder exprBuilder
+	@Inject extension SexecBuilder sexecBuilder
```
Comment (terfloth) [inline plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/transformation/ng/BehaviorMapping.xtend:26]: required?

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/CFlowConfiguration.xtend (hunk 5 above)
Comment (terfloth) [reply plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/CFlowConfiguration.xtend:1]: done

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/StatechartTypes.xtend (hunk 6 above)
Comment (terfloth) [reply plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/StatechartTypes.xtend:164]: done

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/DefaultNamingService.xtend (hunk 7 above)
Comment (terfloth) [reply plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/DefaultNamingService.xtend:280]: done

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/tree/TreeNamingService.xtend (hunk 8 above)
Comment (terfloth) [reply plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/tree/TreeNamingService.xtend:317]: done

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/MethodGenerator.xtend [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+	
+
+	def implementation(Iterable<Method> methods) '''
+	 	«FOR m : methods»
+	 		«m.implementation»
+	 		
+	 	«ENDFOR»
+	 '''
+	
+	def implementation(Method it) '''
```
Comment (rherrmannr) [inline plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/MethodGenerator.xtend:51]: merge master. This has been changed. See https://github.com/Yakindu/statecharts/pull/3135/files

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/StatemachineHeaderFragment.xtend [hunk 11]
```diff
@@ -130,14 +146,9 @@ class StatemachineHeaderFragment implements IHeaderFragment {
 		/*! Initializes the «type» state machine data structures. Must be called before first usage.*/
 		extern «declareInit»
 		
-		/*! Activates the state machine */
-		extern «declareEnter»
 		
-		/*! Deactivates the state machine */
-		extern «declareExit»
```
Comment (rherrmannr) [inline plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/StatemachineHeaderFragment.xtend:137]: missing comment?

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/extensions/SexecBuilder.xtend [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+import org.yakindu.sct.model.sexec.Method
+import org.yakindu.sct.model.sexec.DoWhile
+import org.eclipse.emf.ecore.util.EcoreUtil
+import org.eclipse.emf.ecore.EReference
+import org.yakindu.sct.model.sexec.Call
+import org.yakindu.sct.model.sexec.SexecPackage
+
+class SexecBuilder {
+	
+//	@Inject extension ITypeSystem typeSystem
```
Comment (rherrmannr) [inline plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/extensions/SexecBuilder.xtend:30]: remove comment

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/extensions/SexecBuilder.xtend [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+	@Inject extension TypeBuilder typeBuilder
+	extension SexecFactory sexecFactory = SexecFactory.eINSTANCE
+//	extension TypesFactory typesFactory = TypesFactory.eINSTANCE
```
Comment (rherrmannr) [inline plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/extensions/SexecBuilder.xtend:33]: remove comment

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/extensions/Naming.xtend [hunk 14]
```diff
@@ -141,6 +154,27 @@ class Naming {
...
+	def dispatch String cType(ExecutionFlow it) {
+		type
+	}
+	
+	def dispatch String cType(Type it) {
+		if (! it.targetLanguageName.isNullOrEmpty) return targetLanguageName
+		if (! it.name.isNullOrEmpty) return it.name
+		
+		val typeSuffix = if (it.isEventBuffer) "EvBuf" else ""
+		val typeOrigin = it.originTraces.filter(EObject).findFirst[true]
```
Comment (tkutz) [inline plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/extensions/Naming.xtend:171]: `findFirst[true]` => `head`

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/concepts/EventProcessing.xtend [hunk 15]
```diff
@@ -0,0 +1,382 @@
...
+			
+			if ( ref !== null ) {
+				if (ref.isMany) {
+					val list = parent.eGet(ref) as List<EObject>
+					list.set(list.indexOf(oldValue), newValue)
+				}
+				else parent.eSet(ref, newValue)				
+			}			
+		}	
+	}
```
Comment (tkutz) [inline plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/concepts/EventProcessing.xtend:310]: Why not using `EcoreUtil.replace(o1,o2)` ?

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/concepts/ExecutionGuard.xtend [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+	
+	def _initIsExecuting(ExecutionFlow it) {
+		if (isExecutingProperty !== null) 
+			isExecutingProperty._assign(_false)
+		else
+			_empty
+	}	
+
+	def Property isExecutingProperty(ExecutionFlow it) {
+		features.filter( typeof(Property) ).filter( m | m.name == org.yakindu.sct.model.sexec.concepts.ExecutionGuard.IS_EXECUTING).head
```
Comment (tkutz) [inline plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/concepts/ExecutionGuard.xtend:59]: No need to qualify `IS_EXECUTING` here

File: plugins/org.yakindu.sct.doc.user/src/user-guide/code_generation_c.textile [hunk 17]
```diff
@@ -642,6 +642,30 @@ bc. feature Includes {
 
 ==<!-- End sgen_feature_includes -->==
 
+==<!-- Start sgen_feature_eventprocessing -->==
```
[the remark this answers is not in the document]
Comment (terfloth) [reply plugins/org.yakindu.sct.doc.user/src/user-guide/code_generation_c.textile:645]: is already there ...

File: plugins/org.yakindu.sct.doc.user/src/user-guide/code_generation_cpp.textile [hunk 18]
```diff
@@ -570,6 +570,31 @@ bc. feature API {
...
+    inEventBuffer = false
+    internalEventBuffer = true
+    inEventQueue = false
+}
+
+==</div>==
+
+==<!-- End sgen_feature_eventprocessing -->==
+
+
```
[the remark this answers is not in the document]
Comment (terfloth) [reply plugins/org.yakindu.sct.doc.user/src/user-guide/code_generation_cpp.textile:597]: Already exists at target location - removed from C doc

File: plugins/org.yakindu.sct.doc.user/src/user-guide/code_generation_intro.textile [hunk 19]
```diff
@@ -294,6 +294,30 @@ bc. feature Debug {
...
+bc. feature EventProcessing {
+    inEventBuffer = false
+    internalEventBuffer = true
+    inEventQueue = false
+}
+
+==</div>==
+
+==<!-- End sgen_feature_eventprocessing -->==
+
```
[the remark this answers is not in the document]
Comment (terfloth) [reply plugins/org.yakindu.sct.doc.user/src/user-guide/code_generation_intro.textile:320]: Already exists here

File: plugins/org.yakindu.sct.doc.user/src/user-guide/code_generation_java.textile [hunk 20]
```diff
@@ -990,6 +990,30 @@ bc. feature GeneralFeatures {
...
+bc. feature EventProcessing {
+    inEventBuffer = false
+    internalEventBuffer = true
+    inEventQueue = false
+}
+
+==</div>==
+
+==<!-- End sgen_feature_eventprocessing -->==
+
```
[the remark this answers is not in the document]
Comment (terfloth) [reply plugins/org.yakindu.sct.doc.user/src/user-guide/code_generation_java.textile:1016]: Already exists at target location - removed from Java doc

[the remark this answers is not in the document]
Comment (terfloth) [reply plugins/org.yakindu.sct.doc.user/src/user-guide/code_generation_java.textile:1016]: Applied change.

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/MethodGenerator.xtend (hunk 10 above)
Comment (terfloth) [reply plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/MethodGenerator.xtend:51]: Is not missing - comments are part of method definitions. They are defined by the model sequencer and generated by MethodGenerator.

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/StatemachineHeaderFragment.xtend (hunk 11 above)
Comment (terfloth) [reply plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/StatemachineHeaderFragment.xtend:137]: cleaned up

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/extensions/SexecBuilder.xtend (hunk 12 above)
Comment (terfloth) [reply plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/extensions/SexecBuilder.xtend:30]: cleaned up

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/extensions/Naming.xtend (hunk 14 above)
Comment (terfloth) [reply plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/extensions/Naming.xtend:171]: done

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/concepts/EventProcessing.xtend (hunk 15 above)
Comment (terfloth) [reply plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/concepts/EventProcessing.xtend:310]: It is so nice to do it on my own ;-) 
.... changed

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/concepts/ExecutionGuard.xtend (hunk 16 above)
Comment (terfloth) [reply plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/concepts/ExecutionGuard.xtend:59]: Oh yes - class was moved and xtend does funny things in that case.

changed.

**Pull request merged**